File: 16. 3Sum Closest/Dynamic/3sum_closest.kt
const val number_of_elements: Int = 3

data class Result (var values: Array<Int>, var number_of_values: Int, var not_null: Boolean){
	
	public fun set_values(list: Array<Int>): Unit{
		not_null = true;
		number_of_values = values.size;
		
		for (i in 0 until number_of_values){
			values[i] = list[i];
		}
	}
	
	constructor (list: Array<Int>): this(list, list.size, true){
	}
	
	constructor (): this(Array(number_of_elements){0}, 0, false){
	}
	
	public fun print_values(): Unit{
		println ("This is the hand work of Kotlin");
		var out: Int = 0;
		
		for (i in 0 until number_of_elements){
			print (values[i]);
			out += values[i];
			
			if (i < (number_of_values - 1)){
				print ("\t");
			}
		}
		
		print("\nThe sum of the values is:\t");
		println(out);
	}
	
	public fun empty_result(): Unit{
		not_null = false;
		
		for (i in 0 until number_of_values){
			values[i] = 0;
		}
		
		number_of_values = 0;
	}
}

fun absolute (number: Int) : Int {
	return if (number < 0) (number * -1) else (number * 1);
}

fun sum_of_three_closest (list: Array<Int>, target: Int): MutableList<Result>{
	var closest_difference: Int = 0;
	var first_execution: Boolean = true;
	var len_list: Int = list.size;
	var out: MutableList<Result> = mutableListOf<Result>();
	
	for (a in 0 until len_list){
		for (b in 0 until len_list){
			if (b > a){
				for(c in 0 until len_list){
					if (c > b){
						if ((a != b) && (a != c) && (b != c)){
							var aa: Int = list[a];
							var bb: Int = list[b];
							var cc: Int = list[c];
							
							//if ((aa != bb) && (aa != cc) && (bb != cc)){
							if (true){
								var difference: Int = absolute(target - (aa + bb + cc));
								var update_array: Boolean = false;
								
								if (first_execution){
									closest_difference = difference;
									first_execution = false;
									
									update_array = true;
								} else {
									if (difference < closest_difference){
										closest_difference = difference;
										
										out = mutableListOf<Result>();
										
										update_array = true;
									} else if (difference == closest_difference){
										update_array = true;
									}
								}
								
								if (update_array){
									var y: Array<Int> = arrayOf(aa, bb, cc);
									out.add(Result(y));
								}
							
							}
						}
					}
				}
			}
		}
	}
	return (out);
}

fun main (){
	val nums: Array<Int> = arrayOf(-1,2,1,-4);
	val len_nums: Int = nums.size;
	val target: Int = 1;
	
	//val sample_space = (factorial(len_nums) / (factorial(len_nums - number_of_elements) * factorial(number_of_elements)));
	
	var res: MutableList<Result> = sum_of_three_closest(nums, target);
	
	for (i in 0 until res.size){
		res[i].print_values();
	}
}

File: 16. 3Sum Closest/Dynamic/3sum_closest.rs
const number_of_elements: usize = 3;

#[derive(Copy, Clone)]
struct Result{
	values: [isize; number_of_elements],
	number_of_values: isize,
	not_null: bool
}

impl Result{
	fn new(list: &[isize]) -> Self{
		let len_list = list.len() as isize;
		let mut arr: [isize; number_of_elements] = [0; number_of_elements];
		
		for i in 0..len_list{
			arr[i as usize] = list[i as usize];
		}
		
		return Result{values: arr, number_of_values: len_list, not_null: true};
	}
	
	fn empty() -> Self{
		return Result{values: [0; number_of_elements], number_of_values: number_of_elements as isize, not_null: false};
	}
	
	fn set_values (&mut self, list: &[isize]){
		self.not_null = true;
		self.number_of_values = list.len() as isize;
		
		for i in 0..self.number_of_values{
			self.values[i as usize] = list[i as usize];
		}
	}
	
	fn print_values (self) -> (){
		println!("This is the handy work of Rust");
		
		let mut out: isize = 0;
		
		for i in 0..self.number_of_values{
			print!("{}", self.values[i as usize]);
			out += self.values[i as usize];
			
			if i < (self.number_of_values - 1){
				print!("\t");
			}
		}
		println!("\nThe sum of the values is {}\n\n", out);
	}
	
	fn empty_result (&mut self) -> (){
		self.not_null = false;
		
		for i in 0..self.number_of_values{
			self.values[i as usize] = 0;
		}
		
		self.number_of_values = 0;
	}
}

fn absolute (number: isize) -> isize{
	return if (number < 0) {(number * -1)} else {(number * 1)};
}

fn sum_of_three_closest (list: &[isize], target: isize) -> Vec<Result>{
//fn sum_of_three_closest (list: &[isize], target: isize) -> Result{
	let mut closest_difference: isize = 0;
	let mut first_execution: bool = true;
	let len_list: usize = list.len();
	let mut out: Vec<Result> = vec![];
	//let mut out: Result;
	
	for a in 0..len_list{
		for b in 0..len_list{
			if b > a{
				for c in 0..len_list{
					if c > b{
						if (a != b) && (a != c) && (b != c){
							let aa = list[a];
							let bb = list[b];
							let cc = list[c];
							
							//if (aa != bb) && (aa != cc) && (bb != cc){
							
							if true{	
								let difference: isize = absolute((aa + bb + cc) - target);
								let mut update_array: bool = false;
								
								if first_execution{
									closest_difference = difference;
									first_execution = false;
									update_array = true;
								} else {
									if (difference < closest_difference){
										closest_difference = difference;
										/*
										//This is the one that returns the first combination
										out = Result::empty();	
										*/
									
									//BEGIN if commenting out for the first occurrence outcome
										out = vec![];
										update_array = true;
									} else if (difference == closest_difference){
										update_array = true;
									//END if commenting out for the first occurrence outcome
									}
								}
								
								if update_array{
									let y: [isize; number_of_elements] = [aa, bb, cc];
									
									let res: Result = Result::new(&y);
									
									out.push(res);
								}
							}
						}
					}
				}
			}
		}
	}
	
	return out;
}

fn main (){
	let nums: &[isize] = &[-1,2,1,-4];
	let len_nums: usize = nums.len() as usize;
	
	let res: Vec<Result> = sum_of_three_closest(nums, 1);
	
	for i in 0..res.len(){
		res[i].print_values();
	}
}
